core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -474,72 +454,72 @@ public void run() {
    public CompletionStage<Void> batch(int publisherCount, Publisher<SegmentedPublisher<Object>> removePublisher,
          Publisher<SegmentedPublisher<MarshallableEntry<K, V>>> writePublisher) {
       assertNotStopped();
-      Flowable.fromPublisher(removePublisher)
-            .subscribe(sp ->
-               Flowable.fromPublisher(sp)
-                     .subscribe(key -> submissionFlowable.onNext(new RemoveModification(sp.getSegment(), key)))
-            );
-      Flowable.fromPublisher(writePublisher)
-            .subscribe(sp ->
-                  Flowable.fromPublisher(sp)
-                        .subscribe(me -> submissionFlowable.onNext(new PutModification(sp.getSegment(), me)))
-            );
-      submitBatchIfNecessary();
-      return asyncOrThrottledStage();
-   }
-
-   @Override
-   public CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> entry) {
-      assertNotStopped();
-      submissionFlowable.onNext(new PutModification(segment, entry));
-      submitBatchIfNecessary();
-      return asyncOrThrottledStage();
-   }
-
-   @Override
-   public CompletionStage<Boolean> delete(int segment, Object key) {
-      assertNotStopped();
-      submissionFlowable.onNext(new RemoveModification(segment, key));
-      submitBatchIfNecessary();
-      return asyncOrThrottledStage()
-            // We always assume it was removed with async
-            .thenCompose(ignore -> CompletableFutures.completedTrue());
-   }
-
-   private void submitBatchIfNecessary() {
+      Maybe<CompletionStage<Void>> maybeRemove = Flowable.fromPublisher(removePublisher)
+            .flatMap(sp -> Flowable.fromPublisher(sp)
+                  .map(key -> submitModification(new RemoveModification(sp.getSegment(), key))), publisherCount)
+            .lastElement();
```

Comment:
Yes, as we are invoking `submitModification` in a single thread it guarantees the last invocation will be later than any other invocation in the underlying batched queue. And since all elements in a given batch are completed at the same time, it is okay. I can add a comment though, good call out.

Reply:
That too, but I was also thinking about what happens if one `CompletionStage` completes with an exception and then the last `CompletionStage` completes successfully.

Reply:
Okay.

Reply:
My guess as to why I did it this over 3 months ago was because I didn't know of a great way without having to allocate N more objects for each object where as the chance of getting an error from no user operation was pretty low. But I will add it in now.